File: src/RepositoryModules.VB/ModeExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices
Imports Newtonsoft.Json
Imports Storex

Public Module ModeExtensions

    <Extension>
    Public Function TryExtractProperty(Of T As Class)(Mode As IMode, Key As String, <Out> ByRef Value As T) As Boolean

        Dim objValue As Object = Nothing

        If Mode.ExtendedProperties.TryGetValue(Key, objValue) AndAlso TypeOf objValue Is T Then

            Value = objValue

            Return True

        ElseIf objValue IsNot Nothing Then

            Try
                Value = JsonConvert.DeserializeObject(Of T)(JsonConvert.SerializeObject(objValue))

                Return True

            Catch Ex As JsonException

                Value = Nothing

                Return False
            End Try
        Else

            Value = Nothing

            Return False
        End If
    End Function
End Module

File: src/RepositoryModules.CS/ModeExtensions.cs
using Newtonsoft.Json;
using Storex;

namespace RepositoryModules
{
    public static class ModeExtensions
    {
        public static bool TryExtractProperty<T>(this IMode mode, string key, out T value) where T : class
        {
            if (mode.ExtendedProperties.TryGetValue(key, out var objValue) && objValue is T v)
            {
                value = v;

                return true;
            }
            else if (objValue != null)
            {
                try
                {
                    value = JsonConvert.DeserializeObject<T>(JsonConvert.SerializeObject(objValue));

                    return true;
                }
                catch (JsonException)
                {
                    value = null;

                    return false;
                }
            }
            else
            {
                value = null;

                return false;
            }
        }
    }
}